src/audio/pipeline.c

```diff
@@ -199,7 +202,7 @@ static int pipeline_comp_free(struct comp_dev *current, void *data, int dir)
 	current->pipeline = NULL;
 
 	pipeline_for_each_comp(current, &pipeline_comp_free, data,
-			       NULL, NULL, dir);
+			       NULL, NULL, dir, false);
```

Comment:
@lyakh I could be wrong, but should free also only be done if the components are connected?

Reply:
@ranj063 I can be wrong too, but I don't see anything in `pipeline_comp_complete()`, that would require `.pipeline != NULL`?

Reply:
Sorry, I mean in `pipeline_comp_free()`

Reply:
I wonder what happens if you have a pipeline that is running and you'd do a pipeline_free on an attached component. Do we absolutely know that such pipelines would have different pipeline IDs and would stop propagating anyway?

Reply:
@paulstelian97 I don't know if we can guarantee that, but so far that's exactly how it is - connected pipelines use different IDs

Reply:
If we can't guarantee that we could end up freeing components which are running, so NAK unless there is a way to guarantee it.

Reply:
@paulstelian97 hm? That's exactly as it is now. `false` means keep the previous behaviour. `true` means operate only on a subset - on active elements. This patch can reduce the scope of `pipeline_for_each_comp()`, it doesn't widen it.

Reply:
The patch does extend the behavior because before the active elements and all elements were identical sets. At least that's how I understand it.

Reply:
@paulstelian97 of course it changes behaviour, but not in this case.

Reply:
Either way I'm still concerned. The isolation of pipelines via pipeline ID, which is true at the moment, must be guaranteed in the future as well to keep this function correct.

Reply:
@paulstelian97 sorry, I don't understand you. What do you mean "the isolation of pipelines via pipeline ID is true at the moment?" This patch doesn't change that. It is no less true after this patch as it is before this patch. If it is true now, it is also true after this patch. Whether it is true now - I don't know. I don't know what happens if you create a topology with a fork (e.g. a mixer) and try to assign the same pipeline ID to both branches.

Reply:
Yeah, anyway we must ensure that assigning same pipeline ID isn't possible unless firmware crashes and undefined behavior are allowed.